### PR TITLE
Make DAL not a singleton

### PR DESCRIPTION
**Pull request by clockworksoul**
Makes DAL not a singleton. This had some other minor consequences that needed fixing, but ultimately reduced complexity quite a bit.

Config:

* Uninitialized `config` returns an empty struct instead of nil (so tests can run without explicit configs)

Dataaccess:

* `dal-singleton.go`
  * Renamed to `dal.go`
  * Removed weird state logic and all of the machinery that it required.
  * Handles all of the edge cases I could think of:
    * Successful initialization for a data source will run exactly once per version of a config (but can and will be re-run on a particular source)
    * Failed attempts to initialize will retry with backoff.
    * Change events to the config trigger a re-initialization
    * `Get` is locked during the initialization process.  
    * If it fails to initialize, `Get` returns an error until initialization is successful
    * If the config changes mid-retry, that retry in canceled and a new initialization begins.
    
 * `memory/memory-data-access.go`
   * Now actually DOES use a singleton. Surprise!

Service:

* All references to `dataAccessLayer` have been replaced with a call to `dataaccess.Get`
* All tests now use different names for their test entities to prevent collisions.

All the existing tests run fine and I'm like 90% sure this will work as intended.

### CODE REVIEW

File: service/service.go [hunk 1]
```diff
@@ -213,6 +202,7 @@ func buildLoggingMiddleware(logsous chan RequestEvent) func(http.Handler) http.H
 			userID := "-"
 			tokenString := r.Header.Get("X-Session-Token")
 			if tokenString != "" {
+				dataAccessLayer, _ := dataaccess.Get()
```
Comment (theothertomelliott) [inline service/service.go:205]: If this error isn't handled, could the following line panic?

Comment (clockworksoul) [reply service/service.go:205]: This middle exists entirely for the purposes of logging who did what, and it's entirely possible for a non-nil error to be returned during normal functioning whenever an API request is made without a valid token. The behavior to log a `-` in place of the user name when the user is unknown is correct.

tl;dr -- the middleware shouldn't panic whenever an unknown user tries to interact with the API 🙂

Comment (clockworksoul) [reply service/service.go:205]: **BUT** on further reflection, I can make this more robust!

Comment (clockworksoul) [reply service/service.go]: Okay, ignore everything above because I was chasing the wrong point.

BUT -- I'd still rather not panic because that's hardly ever the helpful thing to do. Instead, the middleware logs an error and returns a status 500. This allows the operator to identify an error rate spike and also to correct a potentially broken config, without messy manics in handler processes.

How's that look? 🙂

File: service/service.go [hunk 2]
```diff
@@ -202,9 +202,15 @@ func buildLoggingMiddleware(logsous chan RequestEvent) func(http.Handler) http.H
...
 			if tokenString != "" {
-				dataAccessLayer, _ := dataaccess.Get()
-				token, _ := dataAccessLayer.TokenRetrieveByToken(r.Context(), tokenString)
-				userID = token.User
+				if dataAccessLayer, err := dataaccess.Get(); err != nil {
+					log.WithError(err).Error(errs.ErrDataAccess)
+					telemetry.Errors().WithError(err).Commit(r.Context())
+					respondAndLogError(r.Context(), w, errs.ErrDataAccess)
+					return
+				} else {
```
Comment (theothertomelliott) [inline service/service.go:210]: Do we need the else block since the above returns?

Comment (clockworksoul) [reply service/service.go]: True! Removed.

**Review by theothertomelliott — APPROVED** [review]
